### ISSUE TITLE
[python-package] Using `lgb.Booster(model_file=model_file)` and `joblib.load()` loading model to predict the same dataset, get different result.

### ISSUE THREAD

**Issue by wzymmzs** [microsoft/LightGBM#5951]
## Description
[python-package] Using `lgb.Booster(model_file=model_file)` and `joblib.load()` loading model to predict the same dataset, get different result.

Using `joblib.load()` can get expectant result. But using `lgb.Booster(model_file=model_file)`  get a result,  recall

## Reproducible example
Not very convenient to display the code. But if any version change code `lgb.Booster(model_file=model_file)` , i can rollback to that version.

## Environment info

LightGBM version or commit hash: python package version 3.3.5

Command(s) you used to install LightGBM

```shell
pip install lightgbm
```

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

This is not enough information for us to be able do more for you than just *guess* at what might be happening.

Can you please share a minimal, reproducible example ([link](https://stackoverflow.com/help/minimal-reproducible-example)), some code we can run which shows exactly what you're trying to do? If you can't share your actual code, that's understandable... you could create a small example using publicly-available data like the datasets from `sklearn.datasets`. 

Here's a good example: https://github.com/microsoft/LightGBM/issues/5617#issuecomment-1338655747

**Comment by wzymmzs** [comment]
[lightgbm demo.zip](https://github.com/microsoft/LightGBM/files/11914789/lightgbm.demo.zip)

I use `model_train` method to get `model.pkl`. Remove unreadable codes get model.txt. Use train data to predict, model.pkl get expectant result. But model.txt get error result.

lightgbm python version: 3.3.5

**Comment by jameslamb** [comment]
I've uploaded the code in plaintext here, and attached only the CSV dataset. In the future, please do that! So that others can see the code without needing to download and unzip a file, and so it'll show up in search results.

```python
from sklearn.metrics import *
from sklearn.model_selection import StratifiedKFold, KFold, train_test_split
import lightgbm as lgb
import pandas as pd
import time
import calendar
import joblib

random_state = 1

def eval(y_test, y_pred, threshold):
    # statistics(y_test, y_pred)

    for i in range(y_pred.__len__()):
        if y_pred[i] > threshold:
            y_pred[i] = 1
        else:
            y_pred[i] = 0

    # eval
    rmse_test = mean_squared_error(y_test, y_pred) ** 0.5
    print(f'The RMSE of prediction is: {rmse_test}')

    # 混淆矩阵
    tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()
    print(f'tn, fp, fn, tp : {tn}, {fp}, {fn}, {tp}')

    # accuracy
    accuracy = accuracy_score(y_test, y_pred)
    print(f'accuracy score: {accuracy}')

    # precision
    precision = precision_score(y_test, y_pred)
    print(f'precision score: {precision}')

    # recall
    recall = recall_score(y_test, y_pred)
    print(f'recall score: {recall}')

    # f1
    f1 = f1_score(y_test, y_pred)
    print(f'f1 score: {f1}')

    # auc
    auc_scoure = roc_auc_score(y_test, y_pred)
    print(f'auc score: {auc_scoure}')


def set_type(df):
    continuous_columns = ['auditor_hired_date', 'auditor_quality_7d', 'auditor_quality_30d',
                          'auditor_valid_time_rate_7d', 'auditor_valid_time_rate_30d',
                          'auditor_pure_time_rate_7d', 'auditor_pure_time_rate_30d',
                          'audit_time', 'video_duration', 'time_level', 'case_duration',
                          'summary_length',
                          'quality_score']

    category_columns = ['auditor_title_id', 'auditor_status',
                        'audit_origin_status', 'content_biz_line', 'biz_code', 'ugc_scene', 'content_type',
                        'is_multi_pics', 'has_item',
                        'grade_machine_value', 'grade_manual_value', 'cert_type', 'creator_g_level', 'creator_role',
                        'is_second_edit']

    df[continuous_columns] = df[continuous_columns].astype('float')
    df[category_columns] = df[category_columns].astype('category')
    print(df.info())
    print(df["check_result"].value_counts())
    return df


def load(file_name):
    df = pd.read_csv(f"target_file/{file_name}.csv", header=0, sep=',', low_memory=False, encoding='utf-8')
    df = df.loc[:, ~df.columns.str.contains('Unnamed')]
    df = set_type(df)
    return df


def spilt_data(df, label_column, random_state):
    df_train, df_test = train_test_split(df, test_size=0.2, stratify=df[label_column], random_state=random_state)
    df_train, df_valid = train_test_split(df_train, test_size=0.2, stratify=df_train[label_column],
                                          random_state=random_state)

    y_train = df_train[label_column]
    y_valid = df_valid[label_column]
    y_test = df_test[label_column]
    x_train = df_train.drop(label_column, axis=1)
    x_valid = df_valid.drop(label_column, axis=1)
    x_test = df_test.drop(label_column, axis=1)
    return y_train, y_valid, y_test, x_train, x_valid, x_test


def model_train(file_name):
    ts = calendar.timegm(time.gmtime())

    print('Starting training...')
    print('Starting loading df...')
    df = load(file_name)

    y_train, y_valid, y_test, x_train, x_valid, x_test = spilt_data(df, "check_result", random_state)

    print('Starting training...')
    gbm = lgb.LGBMClassifier(num_leaves=8,
                             max_depth=3,
                             # min_child_samples=50,
                             learning_rate=0.05,
                             n_estimators=1000,
                             is_unbalance=True)
    gbm.fit(x_train, y_train,
            eval_set=[(x_valid, y_valid)],
            # eval_metric='l2',
            eval_metric='recall',
            # callbacks=[lgb.early_stopping(10000)],
            verbose=100)

    # save model to file
    model_file = f'model/model'
    # gbm.booster_.save_model(model_file + ".txt")
    joblib.dump(gbm, model_file + ".pkl")

    # feature importances
    print(f'Feature importances: {list(gbm.feature_importances_)}')
    print(gbm.feature_name_)

    print('Starting test training data...')
    # predict
    y_pred = gbm.predict(x_train, num_iteration=gbm.best_iteration_)
    eval(y_train, y_pred, 0.9)

    print('Starting predicting...')
    # predict
    y_pred = gbm.predict(x_valid, num_iteration=gbm.best_iteration_)
    eval(y_valid, y_pred, 0.9)

    print('Starting testing...')
    y_pred = gbm.predict(x_test, num_iteration=gbm.booster_.best_iteration)
    eval(y_test, y_pred, 0.9)

    print('Starting testing all...')
    y_all = df["check_result"]
    x_all = df.drop("check_result", axis=1)
    y_pred = gbm.predict(x_all, num_iteration=gbm.booster_.best_iteration)
    eval(y_all, y_pred, 0.9)

    return model_file


def model_test(file_name, model_file, model_type):
    print('Starting testing...')

    if "txt" == model_type:
        bst = lgb.Booster(model_file=f"model/{model_file}")
    elif "pkl" == model_type:
        bst = joblib.load(f"model/{model_file}")

    print('Starting loading df...')
    df = load(file_name)
    y_test = df["check_result"]
    X_test = df.drop("check_result", axis=1)


    # predict
    # print(f'best iteration: {bst.best_iteration}')
    y_pred = bst.predict(X_test)

    print(y_pred)

    eval(y_test, y_pred, 0.9)


if __name__ == "__main__":
    # model_file = model_train("dataset")
    # model_test("dataset", "model.pkl", "pkl")
    model_test("dataset", "model.txt", "txt")

    pass
```

[dataset.csv](https://github.com/microsoft/LightGBM/files/11917085/dataset.csv)

Before we spend time on this, could you please help us some more?

* try to minimize this to the smallest amount of code that reproduces the issue (for example, please replace that `*` import with specific imports, remove unnecessary computation of metrics)
* clarify.... is the problem that you get a "different" result or an "error"? If it's an "error", please provide the exact text of the error message

**Comment by wzymmzs** [comment]
```from sklearn.metrics import confusion_matrix, recall_score
from sklearn.model_selection import train_test_split
import lightgbm as lgb
import pandas as pd
import time
import calendar
import joblib

random_state = 1


def eval(y_test, y_pred, threshold):
    for i in range(y_pred.__len__()):
        if y_pred[i] > threshold:
            y_pred[i] = 1
        else:
            y_pred[i] = 0

    tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()
    print(f'tn, fp, fn, tp : {tn}, {fp}, {fn}, {tp}')

    # recall
    recall = recall_score(y_test, y_pred)
    print(f'recall score: {recall}')


def set_type(df):
    continuous_columns = ['auditor_hired_date', 'auditor_quality_7d', 'auditor_quality_30d',
                          'auditor_valid_time_rate_7d', 'auditor_valid_time_rate_30d',
                          'auditor_pure_time_rate_7d', 'auditor_pure_time_rate_30d',
                          'audit_time', 'video_duration', 'time_level', 'case_duration',
                          'summary_length',
                          'quality_score']

    category_columns = ['auditor_title_id', 'auditor_status',
                        'audit_origin_status', 'content_biz_line', 'biz_code', 'ugc_scene', 'content_type',
                        'is_multi_pics', 'has_item',
                        'grade_machine_value', 'grade_manual_value', 'cert_type', 'creator_g_level', 'creator_role',
                        'is_second_edit']

    df[continuous_columns] = df[continuous_columns].astype('float')
    df[category_columns] = df[category_columns].astype('category')
    print(df.info())
    print(df["check_result"].value_counts())
    return df


def load(file_name):
    df = pd.read_csv(f"target_file/{file_name}.csv", header=0, sep=',', low_memory=False, encoding='utf-8')
    df = df.loc[:, ~df.columns.str.contains('Unnamed')]
    df = set_type(df)
    return df


def spilt_data(df, label_column, random_state):
    df_train, df_test = train_test_split(df, test_size=0.2, stratify=df[label_column], random_state=random_state)
    df_train, df_valid = train_test_split(df_train, test_size=0.2, stratify=df_train[label_column],
                                          random_state=random_state)

    y_train = df_train[label_column]
    y_valid = df_valid[label_column]
    y_test = df_test[label_column]
    x_train = df_train.drop(label_column, axis=1)
    x_valid = df_valid.drop(label_column, axis=1)
    x_test = df_test.drop(label_column, axis=1)
    return y_train, y_valid, y_test, x_train, x_valid, x_test


def model_train(file_name):
    ts = calendar.timegm(time.gmtime())

    print('Starting training...')
    print('Starting loading df...')
    df = load(file_name)

    y_train, y_valid, y_test, x_train, x_valid, x_test = spilt_data(df, "check_result", random_state)

    print('Starting training...')
    gbm = lgb.LGBMClassifier(num_leaves=8,
                             max_depth=3,
                             # min_child_samples=50,
                             learning_rate=0.05,
                             n_estimators=1000,
                             is_unbalance=True)
    gbm.fit(x_train, y_train,
            eval_set=[(x_valid, y_valid)],
            # eval_metric='l2',
            eval_metric='recall',
            # callbacks=[lgb.early_stopping(10000)],
            verbose=100)

    # save model to file
    model_file = f'model/model'
    # gbm.booster_.save_model(model_file + ".txt")
    joblib.dump(gbm, model_file + ".pkl")

    # feature importances
    print(f'Feature importances: {list(gbm.feature_importances_)}')
    print(gbm.feature_name_)

    print('Starting testing all...')
    y_all = df["check_result"]
    x_all = df.drop("check_result", axis=1)
    y_pred = gbm.predict(x_all, num_iteration=gbm.booster_.best_iteration)
    eval(y_all, y_pred, 0.9)

    return model_file


def model_test(file_name, model_file, model_type):
    print('Starting testing...')

    if "txt" == model_type:
        bst = lgb.Booster(model_file=f"model/{model_file}")
    elif "pkl" == model_type:
        bst = joblib.load(f"model/{model_file}")

    print('Starting loading df...')
    df = load(file_name)
    y_test = df["check_result"]
    X_test = df.drop("check_result", axis=1)

    # predict
    y_pred = bst.predict(X_test)

    print(y_pred)

    eval(y_test, y_pred, 0.9)


if __name__ == "__main__":
    model_file = model_train("dataset")
    # model_test("dataset", "model.pkl", "pkl")
    # model_test("dataset", "model.txt", "txt")

    pass

```

Using complete data to test has representativeness. Method `model_file = model_train("dataset")` get result:
```
tn, fp, fn, tp : 45458, 20051, 2540, 8507
recall score: 0.7700733230741378
```

Method `model_test("dataset", "model.pkl", "pkl")` get result:
```
tn, fp, fn, tp : 45458, 20051, 2540, 8507
recall score: 0.7700733230741378
```

Method `model_test("dataset", "model.txt", "txt")` get result:
```
tn, fp, fn, tp : 65490, 19, 11007, 40
recall score: 0.003620892550013578
```

We can see that, 'confusion_matrix' is different from the result by different type of model file. Format of 'pkl' is correct, format of 'txt' is incorrect. 
Although not all datasets will yield different results, there is a problem with this case.

**Comment by wzymmzs** [comment]
Additionally, sometimes the model results have a probability of 0 to 1, and sometimes only 0 and 1 are the two results. Is there any configuration can specify ?

**Comment by wzymmzs** [comment]
> ```
> from sklearn.model_selection import train_test_split
> import lightgbm as lgb
> import pandas as pd
> import time
> import calendar
> import joblib
> 
> random_state = 1
> 
> 
> def eval(y_test, y_pred, threshold):
>     for i in range(y_pred.__len__()):
>         if y_pred[i] > threshold:
>             y_pred[i] = 1
>         else:
>             y_pred[i] = 0
> 
>     tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()
>     print(f'tn, fp, fn, tp : {tn}, {fp}, {fn}, {tp}')
> 
>     # recall
>     recall = recall_score(y_test, y_pred)
>     print(f'recall score: {recall}')
> 
> 
> def set_type(df):
>     continuous_columns = ['auditor_hired_date', 'auditor_quality_7d', 'auditor_quality_30d',
>                           'auditor_valid_time_rate_7d', 'auditor_valid_time_rate_30d',
>                           'auditor_pure_time_rate_7d', 'auditor_pure_time_rate_30d',
>                           'audit_time', 'video_duration', 'time_level', 'case_duration',
>                           'summary_length',
>                           'quality_score']
> 
>     category_columns = ['auditor_title_id', 'auditor_status',
>                         'audit_origin_status', 'content_biz_line', 'biz_code', 'ugc_scene', 'content_type',
>                         'is_multi_pics', 'has_item',
>                         'grade_machine_value', 'grade_manual_value', 'cert_type', 'creator_g_level', 'creator_role',
>                         'is_second_edit']
> 
>     df[continuous_columns] = df[continuous_columns].astype('float')
>     df[category_columns] = df[category_columns].astype('category')
>     print(df.info())
>     print(df["check_result"].value_counts())
>     return df
> 
> 
> def load(file_name):
>     df = pd.read_csv(f"target_file/{file_name}.csv", header=0, sep=',', low_memory=False, encoding='utf-8')
>     df = df.loc[:, ~df.columns.str.contains('Unnamed')]
>     df = set_type(df)
>     return df
> 
> 
> def spilt_data(df, label_column, random_state):
>     df_train, df_test = train_test_split(df, test_size=0.2, stratify=df[label_column], random_state=random_state)
>     df_train, df_valid = train_test_split(df_train, test_size=0.2, stratify=df_train[label_column],
>                                           random_state=random_state)
> 
>     y_train = df_train[label_column]
>     y_valid = df_valid[label_column]
>     y_test = df_test[label_column]
>     x_train = df_train.drop(label_column, axis=1)
>     x_valid = df_valid.drop(label_column, axis=1)
>     x_test = df_test.drop(label_column, axis=1)
>     return y_train, y_valid, y_test, x_train, x_valid, x_test
> 
> 
> def model_train(file_name):
>     ts = calendar.timegm(time.gmtime())
> 
>     print('Starting training...')
>     print('Starting loading df...')
>     df = load(file_name)
> 
>     y_train, y_valid, y_test, x_train, x_valid, x_test = spilt_data(df, "check_result", random_state)
> 
>     print('Starting training...')
>     gbm = lgb.LGBMClassifier(num_leaves=8,
>                              max_depth=3,
>                              # min_child_samples=50,
>                              learning_rate=0.05,
>                              n_estimators=1000,
>                              is_unbalance=True)
>     gbm.fit(x_train, y_train,
>             eval_set=[(x_valid, y_valid)],
>             # eval_metric='l2',
>             eval_metric='recall',
>             # callbacks=[lgb.early_stopping(10000)],
>             verbose=100)
> 
>     # save model to file
>     model_file = f'model/model'
>     # gbm.booster_.save_model(model_file + ".txt")
>     joblib.dump(gbm, model_file + ".pkl")
> 
>     # feature importances
>     print(f'Feature importances: {list(gbm.feature_importances_)}')
>     print(gbm.feature_name_)
> 
>     print('Starting testing all...')
>     y_all = df["check_result"]
>     x_all = df.drop("check_result", axis=1)
>     y_pred = gbm.predict(x_all, num_iteration=gbm.booster_.best_iteration)
>     eval(y_all, y_pred, 0.9)
> 
>     return model_file
> 
> 
> def model_test(file_name, model_file, model_type):
>     print('Starting testing...')
> 
>     if "txt" == model_type:
>         bst = lgb.Booster(model_file=f"model/{model_file}")
>     elif "pkl" == model_type:
>         bst = joblib.load(f"model/{model_file}")
> 
>     print('Starting loading df...')
>     df = load(file_name)
>     y_test = df["check_result"]
>     X_test = df.drop("check_result", axis=1)
> 
>     # predict
>     y_pred = bst.predict(X_test)
> 
>     print(y_pred)
> 
>     eval(y_test, y_pred, 0.9)
> 
> 
> if __name__ == "__main__":
>     model_file = model_train("dataset")
>     # model_test("dataset", "model.pkl", "pkl")
>     # model_test("dataset", "model.txt", "txt")
> 
>     pass
> ```
> 
> Using complete data to test has representativeness. Method `model_file = model_train("dataset")` get result:
> 
> ```
> tn, fp, fn, tp : 45458, 20051, 2540, 8507
> recall score: 0.7700733230741378
> ```
> 
> Method `model_test("dataset", "model.pkl", "pkl")` get result:
> 
> ```
> tn, fp, fn, tp : 45458, 20051, 2540, 8507
> recall score: 0.7700733230741378
> ```
> 
> Method `model_test("dataset", "model.txt", "txt")` get result:
> 
> ```
> tn, fp, fn, tp : 65490, 19, 11007, 40
> recall score: 0.003620892550013578
> ```
> 
> We can see that, 'confusion_matrix' is different from the result by different type of model file. Format of 'pkl' is correct, format of 'txt' is incorrect. Although not all datasets will yield different results, there is a problem with this case.

I need results from the 'txt' model to be right. Because I need convert it to 'pmml' file, so then get rest interface.

**Comment by jameslamb** [comment]
Thanks for that. This is still quite a lot of code.

Can you please try to reduce it to a **minimal** example showing the issue you're reporting?

* no commented-out code
* no `if __name___ == "__main__"` block (just assume the whole thing will be run as a script)
* no unnecessary statements like `ts = calendar.timegm(time.gmtime())` or corresponding unnecessary
* `n_estimators` reduced to a smaller value like `10` (assuming that still reproduces the issue)
* just use something like `numpy.testing.assert_array_equal()` or similar instead of that custom `eval()` function to show that the predictions are different

Doing the will reduce the effort required to identify the exact cause of the issue you're reporting.

**Comment by wzymmzs** [comment]
```
import lightgbm as lgb
import pandas as pd
import numpy as np
import joblib


def set_type(df):
    continuous_columns = ['auditor_hired_date', 'auditor_quality_7d', 'auditor_quality_30d',
                          'auditor_valid_time_rate_7d', 'auditor_valid_time_rate_30d',
                          'auditor_pure_time_rate_7d', 'auditor_pure_time_rate_30d',
                          'audit_time', 'video_duration', 'time_level', 'case_duration',
                          'summary_length',
                          'quality_score']

    category_columns = ['auditor_title_id', 'auditor_status',
                        'audit_origin_status', 'content_biz_line', 'biz_code', 'ugc_scene', 'content_type',
                        'is_multi_pics', 'has_item',
                        'grade_machine_value', 'grade_manual_value', 'cert_type', 'creator_g_level', 'creator_role',
                        'is_second_edit']

    df[continuous_columns] = df[continuous_columns].astype('float')
    df[category_columns] = df[category_columns].astype('category')
    return df


def load(file_name):
    df = pd.read_csv(f"target_file/{file_name}.csv", header=0, sep=',', low_memory=False, encoding='utf-8')
    df = df.loc[:, ~df.columns.str.contains('Unnamed')]
    df = set_type(df)
    return df


def model_test(file_name, model_file, model_type):
    print('Starting testing...')

    if "txt" == model_type:
        bst = lgb.Booster(model_file=f"model/{model_file}")
    elif "pkl" == model_type:
        bst = joblib.load(f"model/{model_file}")

    print('Starting loading df...')
    df = load(file_name)
    X_test = df.drop("check_result", axis=1)

    # predict
    y_pred = bst.predict(X_test)

    return y_pred


pkl_result = model_test("dataset", "model.pkl", "pkl")
print(pkl_result)
txt_result = model_test("dataset", "model.txt", "txt")
print(txt_result)
np.testing.assert_array_equal(pkl_result, txt_result)
```

Using the `bst.predict` method. 'pkl' model get 0 and 1. 'txt' model get 0 to 1. I didn't use 'predict_proba' method.

**Comment by jameslamb** [comment]
> 'pkl' model get 0 and 1. 'txt' model get 0 to 1.

Oh I see! Thanks for trying to minimize the example so it's easier to understand what you're reporting. What you've provided still isn't a reproducible example (you've now omitted the code for training models, so the code in https://github.com/microsoft/LightGBM/issues/5951#issuecomment-1617180504 wouldn't work by itself), but I think I understand.

The difference you're seeing is because what you're storing to text and what you're storing with `joblib` are different objects.

What you've stored in a text file is a `lightgbm.Booster` object

What you've stored with `joblib` is an instance of `lightgbm.sklearn.LGBMClassifier`, which wraps a `lightgbm.Booster` in a `scikit-learn`-compatible interface.

By default for binary classification, `LGBMClassifier.predict()` returns predicted classes (to be consistent with the guidance at https://scikit-learn.org/stable/developers/develop.html).

`lightgbm.Booster.predict()`, on the other hand, defaults to returning probabilities of each case being the positive class (`y=1`).

See the example code below, which will work with both the latest released version of `lightgbm` (v3.3.5) and the latest development version of `lightgbm` here on GitHub.

```python
import lightgbm as lgb
import numpy as np
from sklearn.datasets import load_breast_cancer

X, y = load_breast_cancer(return_X_y=True)

model = lgb.LGBMClassifier(n_estimators=10, min_child_samples=1)
model.fit(X, y)

# LGBMClassifier.predict() returns classes by default
sklearn_preds = model.predict(X)
sklearn_preds[:20]
# array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1])

# Booster.predict() returns probability of y_i=1
booster_preds = model.booster_.predict(X)
# array([0.22205477, 0.22205477, 0.22205477, 0.22205477, 0.22205477,
#        0.22205477, 0.22205477, 0.22205477, 0.22205477, 0.22205477,
#        0.22205477, 0.22205477, 0.22205477, 0.22205477, 0.22205477,
#        0.22205477, 0.22205477, 0.22205477, 0.22205477, 0.86518328])

# confirm that Booster.predict() produces probabilities
sklearn_probabilities = model.predict_proba(X)[:,1]
np.testing.assert_array_equal(booster_preds, sklearn_probabilities)

# confirm that Booster.predict()'s probabilities exactly correspond to
# LGBMClassifier.predict()'s classes
booster_classes = np.argmax(
    a=model.predict_proba(X),
    axis=1
)
np.testing.assert_array_equal(sklearn_preds, booster_classes)
```

> Because I need convert it to 'pmml' file

There isn't an official LightGBM-maintained converter for PMML in this project currently. You could try these projects:

* https://github.com/jpmml/sklearn2pmml
* https://github.com/jpmml/jpmml-sklearn
* https://github.com/jpmml/jpmml-lightgbm

I haven't personally used those so can't say for sure if they'll work well for you, but that's where I'd start if I wanted to convert a LightGBM model to PMML.

**Comment by wzymmzs** [comment]
Thanks for explaining my problem.
